extensions/iceberg/src/main/java/io/deephaven/iceberg/util/SchemaSpec.java

```diff
@@ -0,0 +1,140 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.iceberg.util;
+
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+
+import java.util.Iterator;
+
+/**
+ * A specification for extracting the schema from a table.
+ */
+interface SchemaSpec {
+
+    /**
+     * Returns the schema for the given table based on this SchemaSpec.
+     */
+    Schema getSchema(Table table);
```

Comment:
I'd like to make this an implementation detail; end-users should not need to use this. I would suggest either an internal interface or abstract class that defines this, and then our implementation can do:

```java
this.userSchema = ((SchemaSpecImpl)tableWriterOptions.schemaSpec()).getSchema(table);
```